cmd/fs-v1-metadata.go

```diff
@@ -274,167 +271,119 @@ func newFSMetaV1() (fsMeta fsMetaV1) {
-
-		// Entries found shouldn't be more than total
-		// expected config directories, files.
-		if len(entriesSet) > len(expectedConfigsSet) {
-			return errCorruptedFormat
-		}
-
-		// Look for the difference between entries and the
-		// expected config set, resulting entries if they
-		// intersect with original entries set we know
-		// that the backend has unexpected files.
-		if !entriesSet.Difference(expectedConfigsSet).IsEmpty() {
-			return errCorruptedFormat
-		}
+	// No data on disk `format.json` still empty treat
+	// it as unformatted disk.
+	if errorCause(err) == io.EOF {
+		return errUnformattedDisk
 	}
-	return nil
-}
 
-// Check for sanity of FS format for a given version.
-func checkFormatSanityFS(fsPath string, fsFormatVersion string) (err error) {
-	switch fsFormatVersion {
-	case fsFormatV2:
-		err = checkFormatSanityFSV2(fsPath)
-	default:
-		err = errCorruptedFormat
-	}
-	return err
+	return nil
 }
 
-// Initializes a new `format.json` if not present, validates `format.json`
-// if already present and migrates to newer version if necessary. Returns
-// the final format version.
-func initFormatFS(fsPath, fsUUID string) (err error) {
-	fsFormatPath := pathJoin(fsPath, minioMetaBucket, fsFormatJSONFile)
-
-	// fsFormatJSONFile - format.json file stored in minioMetaBucket(.minio.sys) directory.
-	lk, err := lock.LockedOpenFile(preparePath(fsFormatPath), os.O_RDWR|os.O_CREATE, 0600)
+// Check if disk has already a valid format.
+func checkValidFormatFS(fsFormatPath string) error {
+	rlk, err := lock.RLockedOpenFile(preparePath(fsFormatPath))
 	if err != nil {
+		if os.IsNotExist(err) {
+			// If format.json not found then
+			// its an unformatted disk.
+			return traceError(errUnformattedDisk)
+		}
 		return traceError(err)
 	}
-	defer lk.Close()
+	defer rlk.Close()
 
 	var format = &formatConfigV1{}
-	_, err = format.ReadFrom(lk)
-	// For all unexpected errors, we return.
-	if err != nil && errorCause(err) != io.EOF {
-		return traceError(fmt.Errorf("Unable to load 'format.json', %s", err))
+	if err = loadFormatFS(format, rlk.LockedFile); err != nil {
+		return traceError(err)
 	}
+	return checkFormatFS(format)
+}
 
-	// If we couldn't read anything, The disk is unformatted.
-	if errorCause(err) == io.EOF {
-		err = errUnformattedDisk
-		format = newFSFormatV2()
-	} else {
-		// Validate loaded `format.json`.
-		err = checkFormatFS(format, fsFormatVersion)
-		if err != nil && err != errFSFormatOld {
-			return traceError(fmt.Errorf("Unable to validate 'format.json', %s", err))
-		}
-	}
+func initFormatFS(fsPath, fsUUID string) error {
+	fsFormatPath := pathJoin(fsPath, minioMetaBucket, fsFormatJSONFile)
 
-	// Disk is in old format migrate object metadata.
-	if err == errFSFormatOld {
-		if merr := migrateFSObject(fsPath, fsUUID); merr != nil {
-			return merr
+	for {
+		err := checkValidFormatFS(fsFormatPath)
+		if err != nil && errorCause(err) != errUnformattedDisk {
```

Comment:
Could move `return err` to `else if err != nil` clause in line 380 (saves one  `errorCause(err) == errUnformattedDisk` comparison)